src/helpers/ring.rs

```diff
@@ -0,0 +1,201 @@
+}
+
+/// Trait for MPC helpers to communicate with each other. Helpers can send messages and
+/// receive messages from a specific helper.
+#[async_trait]
+pub trait Ring {
+    /// Send message to the destination. Implementations are free to choose whether it is required
+    /// to wait until `dest` acknowledges message or simply put it to a outgoing queue
+    async fn send<T: Message>(&self, dest: HelperAddr, msg: T) -> Result<(), Error>;
+    async fn receive<T: Message>(&self, source: HelperAddr) -> Result<T, Error>;
+}
+
+#[cfg(test)]
+pub mod mock {
+    use crate::helpers::error::Error;
+    use crate::helpers::ring::{HelperAddr, Message, Ring};
+    use async_trait::async_trait;
+    use std::any::TypeId;
+    use std::collections::hash_map::Entry;
+    use std::collections::HashMap;
+    use std::sync::{Arc, Mutex};
+    use tokio::sync::mpsc::{channel, Sender};
+
+    /// Internally we represent all messages to be a sequence of bytes and store them inside
+    /// a hashmap where each element is addressable by message type id and destination (i.e. who
+    /// is the intended receiver of this message).
+    type MessageBuf = HashMap<(HelperAddr, TypeId), Box<[u8]>>;
+
+    /// Each message is packed inside an envelope with some meta information about it.
+    #[derive(Debug)]
+    struct MessageEnvelope {
+        source: HelperAddr,
+        type_id: TypeId,
+        payload: Box<[u8]>,
+    }
+
+    /// A mock implementation of `Ring` trait to be used in unit tests where all helpers are running
+    /// inside the same process. Provides simple and inefficient implementation by buffering all messages
+    /// on `send` and polling the buffer on `receive`. Message is determined to be the same if it has
+    /// the same `TypeId` and helper address matches the destination. For example, message `Foo` sent
+    /// to Helper 2 will be received and removed from the local buffer only when Helper 2 attempts to
+    /// receive it.
+    #[derive(Debug)]
+    pub struct TestHelper {
+        // A handle to send message to this helper
+        input_queue: Sender<MessageEnvelope>,
+
+        // Reference to helper channel on the left side
+        left: Option<Sender<MessageEnvelope>>,
+
+        // Reference to helper channel on the right side
+        right: Option<Sender<MessageEnvelope>>,
+
+        // buffer for messages sent to this helper
+        buf: Arc<Mutex<MessageBuf>>,
+    }
+
+    impl TestHelper {
+        /// Constructs a new instance of test helper using the specified `buf_capacity` buffer
+        /// capacity for the internally used channel.
+        ///
+        /// ## Panics
+        /// Panics if Mutex used internally for synchronization is poisoned or if there are more
+        /// than one message with the same type id and destination address arriving via `send` call.
+        #[must_use]
+        pub fn new(buf_capacity: usize) -> Self {
+            let (tx, mut rx) = channel::<MessageEnvelope>(buf_capacity);
```

Comment:
What do "tx" and "rx" mean? These variable names aren't super obvious in their meaning (to me)